### PR TITLE
switch plugins to use AutoService

### PR DESCRIPTION
**Pull request by arifogel**
This change is [Reviewable](https://reviewable.io/reviews/batfish/batfish/603)

### CODE REVIEW

File: projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java [hunk 1]
```diff
@@ -150,153 +124,12 @@ private boolean isJavaSerializationData(PushbackInputStream stream) throws IOExc
...
-            throw new BatfishException(
-                "Error loading plugin folder: '" + pluginDir.toString() + "'", e);
-          }
-        }
-      }
+    Set<Plugin> plugins =
+        new LinkedHashSet<>(
+            Lists.newArrayList(ServiceLoader.load(Plugin.class, _currentClassLoader)));
+    for (Plugin plugin : plugins) {
+      plugin.initialize(this);
```
Comment (dhalperi) [inline projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java]: I'd probably wrap this in a try/catch and report an error loading any of these but continue loading others?

Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java]: Hmm I don't know. I don't think it really makes sense to keep running when a plugin has failed. How about catch, wrap, and throw?

**Comment by arifogel** [comment]
Rebased and resolved conflict with OutliersQuestionPlugin.java (added AutoService annotation, imports on top of master version)

File: projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java (hunk 1 above)
Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java]: Addressed via catching of ServiceConfigurationException, throwing CompositeBatfishException after all are done loading.

**Comment by dhalperi** [comment]
Reviewed 70 of 71 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/batfish, line 5 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJyRUMg99x3gCNa0oo:-KxJyRUMg99x3gCNa0op:bnw6bsb) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish/batfish#L5)):*
> ```
> BATFISH_PATH="$(dirname $BATFISH)"
> ALLINONE_PATH="${BATFISH_PATH}/../allinone"
> ALLINONE_JAR="$ALLINONE_PATH/target/allinone-bundle-${BATFISH_VERSION}.jar"
> ```

What's the process you use to decide to include `${X}` vs just `$X`?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java, line 8 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJzR2dBU7IA40FssnR:-KxJzR2dBU7IA40FssnS:b2xbo1f) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java#L8)):*
> ```Java
> 
>   @Override
>   public final int compareTo(Plugin o) {
> ```

Should this really be `final`? Would plugin equality not need to be tested later for extensions that care about other fields?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/603)*

**Comment by arifogel** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/batfish, line 5 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJyRUMg99x3gCNa0oo:-KxKMCI2XqQRFknz9_69:bs3an9e) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish/batfish#L5)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

What's the process you use to decide to include `${X}` vs just `$X`?
</blockquote></details>

Really it is only necessary when there is a separator character (e.g. "-") in the name. But I like to use it all the time. Sometimes I am weak and forget.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java, line 8 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJzR2dBU7IA40FssnR:-KxKMTjLw8zbKOFF-r5A:b-iwnwci) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java#L8)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Should this really be `final`? Would plugin equality not need to be tested later for extensions that care about other fields?
</blockquote></details>

The hope is no. Can you provide a plausible counterexample?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/603)*

**Comment by dhalperi** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java, line 8 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJzR2dBU7IA40FssnR:-KxKNlXImf6LFJE3lkhQ:b-sg0mv4) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java#L8)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

The hope is no. Can you provide a plausible counterexample?
</blockquote></details>

Perhaps a better question would be why these need to be comparable at all; what's this actually used for?

If it's for putting them in some sorted order, probably makes more sense to have a delegate comparator of the names than to actually make `Plugin` itself comparable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/603)*

**Comment by arifogel** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java, line 8 at r2](https://reviewable.io:443/reviews/batfish/batfish/603#-KxJzR2dBU7IA40FssnR:-KxKPwwyp9iTh2iAD45z:b-1und3v) ([raw file](https://github.com/batfish/batfish/blob/1931d897a51b9bc8435eedc1126a67805bfbcaa6/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/Plugin.java#L8)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Perhaps a better question would be why these need to be comparable at all; what's this actually used for?

If it's for putting them in some sorted order, probably makes more sense to have a delegate comparator of the names than to actually make `Plugin` itself comparable.
</blockquote></details>

Yes it is just for sorting. From my perspective, if it is ok for `equals` and `hashCode` being defined as they are and final, then it is not a jump to accept `compareTo` existing as implemented and being final. If someone wants to sort in a different order or using different fields, they can use a different comparator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/603)*

**Pull request merged**